xmppserver/src/main/java/org/jivesoftware/openfire/component/InternalComponentManager.java

```diff
@@ -708,29 +708,35 @@ public void joinedCluster( final byte[] nodeID )
     @Override
     public void leftCluster()
     {
-        // Upon leaving a cluster, the server uses its non-clustered/default ID again. Here, all clustered IDs are replaced with the new identity.
-        final NodeID defaultNodeID = XMPPServer.getInstance().getDefaultNodeID();
-        final NodeID nodeID = XMPPServer.getInstance().getNodeID();
-        if ( !defaultNodeID.equals( nodeID ) ) // In more recent versions of Openfire, the ID does not change.
-        {
-            CacheUtil.replaceValueInMultivaluedCache( componentCache, nodeID, defaultNodeID );
+        // The local cluster node left the cluster.
+        if (XMPPServer.getInstance().isShuttingDown()) {
+            // Do not put effort in restoring the correct state if we're shutting down anyway.
+            return;
         }
 
-        // The local cluster node left the cluster.
-        //
+        // Upon leaving a cluster, clustered caches are reset to their local equivalent (by the swap from the clustered
+        // cache implementation to the default cache implementation that's done in the implementation of
+        // org.jivesoftware.util.cache.CacheFactory.leftCluster). This means that they now hold no data (as a new cache
```

Comment:
I don't doubt what you've written here ("This means that they now hold no data") but this seems ... surprising. There's a fair bit of logic in `CacheFactory#leftCluster()` that seems to be redundant and/or straight up broken. A comment to that effect in there may be useful.

Reply:
I do not understand the point that you're trying to make.

As I read the code, this is the event of things when the local node leaves the cluster:

1. `CacheFactory#leftCluster()` gets invoked, which swaps the implementation of each Cache. This swaps the factory for caches for one that provided `DefaultCache` (which is non-clustered), and replaces the wrapped cache implemention, by removing the Hazelcast-backed version with a new `DefaultCache` implementation. As the cache instance is new, it is empty.
2. All `org.jivesoftware.openfire.cluster.ClusterEventListener#leftCluster()` event listeners are invoked. This invokes the method to which you added this comment. As the cache is now empty, it will need to be repopulated with (local) data. This is done by the `restoreCacheContent()` invocation. After the cache is repopulated, any and all event listeners that need to be triggered to reflect the fact that we've "lost" resources on the cluster are invoked

That last bit, triggering the event listeners, can be tricky, as it can be hard to figure out what data was 'lost' (as it's already gone).

Reply:
Sorry, was probably being a bit obtuse. 

My basic point was that I was surprised that open leaving a cluster, a clustered cache is completely emptied. A comment, somewhere, to that effect would be useful, perhaps in CacheFactory.

In more detail, looking in `CacheFactory#leftCluster()` - https://github.com/igniterealtime/Openfire/blob/master/xmppserver/src/main/java/org/jivesoftware/util/cache/CacheFactory.java#L904 - line 904 seems to attempt to put the old entries of the cache back in to the newly wrapped cache - `standaloneCache.putAll(cache);`. If the newly wrapped cache is always empty then I can only assume `cache` is always empty by this point. This looks like a bug to me, but not sure if it's even possible to fix, or if it's wise to do so.

Reply:
Hah! That `putAll()` is the fix for OF-974, which is _exactly_ what is being reverted by the first commit in this PR!

I'm assuming a `putAll()` is used as a result of re-using the same construct that was used when _joining_ a cluster (in [CacheFactory#joinedCluster()](https://github.com/igniterealtime/Openfire/blob/master/xmppserver/src/main/java/org/jivesoftware/util/cache/CacheFactory.java#L882)). There, the cache that's being operated on is not empty: It "contains" all data that's available on all other cluster nodes (through Hazelcast).

Reply:
This basically is the crux of issue that I'm trying to solve. Those `putAll()`'s are effective, but to crude.

A notable problem is when there are cache entries both on the cluster as on the local node that both use the same key, but different values. With the `putAll()`, one of them will be lost. That is the cause of OF-2060.

Reply:
Doh! Missed that. Hmm, this might be a better case for implementing a new listener method somewhere - the method takes the existing set of data, and throws away the stuff it doesn't want, firing off any listeners. The default implementation throws away everything.

Reply:
Yeah, I was thinking of that. My only concern is that this would require the code to pull in all cached data to a cluster node that has already left the cluster - which won't work, as it's no longer in the cluster. The obvious alternative being a setup in wich all data is being stored on all cluster nodes, which probably is unacceptable from a scalability perspective.

Maybe it's feasible to define a smaller data object, that only contains enough data needed to invoke event listeners et al? Maaaaybe all _key_ values of each cache is always available on all cluster nodes, or something like that?

Reply:
Though on reflection - this code is only triggered when the local node leaves the cluster. This is probably an exceptional condition anyway (split brain, or cluster turned off manually are the only two cases I can think of).

So either accept (a) listeners aren't fired - i.e. as implemented in this PR, or (b) only 'some' listeners are fired if the number of nodes in the cluster before leaving was > 2.

But if this PR stands as-is, I would like to see somewhere a nice big comment highlighting that plugins that make use of  their own clustered caches will need to realise that the cache will be empty when the local node leaves a cluster.

Reply:
It's already being mentioned in https://github.com/igniterealtime/Openfire/blob/master/xmppserver/src/main/java/org/jivesoftware/openfire/cluster/ClusterEventListener.java#L34-L36 and https://github.com/igniterealtime/Openfire/blob/master/xmppserver/src/main/java/org/jivesoftware/openfire/cluster/ClusterEventListener.java#L58-L59

If we can make things more clear, I'm all for it though. Should we add it to the Javadoc of Cache itself?

Reply:
I've added a bit of additional javadoc to the Cache interface.